processor/normalizesumsprocessor/processor.go

```diff
@@ -97,229 +93,96 @@ func (nsp *NormalizeSumsProcessor) transformMetrics(rms pdata.ResourceMetrics) [
-	dps := metric.DoubleSum().DataPoints()
+	dps := metric.Sum().DataPoints()
 
 	// Only transform data when the StartTimestamp was not set
-	if dps.Len() > 0 && dps.At(0).StartTimestamp() == 0 {
-		for i := 0; i < dps.Len(); {
-			reportData := nsp.processDoubleSumDataPoint(dps.At(i), resource, metric)
-
-			if !reportData {
-				removeDoubleDataPointAt(dps, i)
-				continue
-			}
-			i++
-		}
+	if dps.Len() == 0 || dps.At(0).StartTimestamp() != 0 {
+		return true, nil
 	}
 
-	return dps.Len()
-}
-
-func (nsp *NormalizeSumsProcessor) processIntSumMetric(resource pdata.Resource, metric pdata.Metric) int {
-	dps := metric.IntSum().DataPoints()
-
-	// Only transform data when the StartTimestamp was not set
-	if dps.Len() > 0 && dps.At(0).StartTimestamp() == 0 {
-		for i := 0; i < dps.Len(); {
-			reportData := nsp.processIntSumDataPoint(dps.At(i), resource, metric)
+	out := pdata.NewNumberDataPointSlice()
+	out.EnsureCapacity(dps.Len())
 
-			if !reportData {
-				removeIntDataPointAt(dps, i)
-				continue
-			}
-			i++
-		}
+	for i := 0; i < dps.Len(); i++ {
+		nsp.processSumDataPoint(dps.At(i), resource, metric, out)
 	}
 
-	return dps.Len()
-}
-
-func (nsp *NormalizeSumsProcessor) processDoubleSumDataPoint(dp pdata.DoubleDataPoint, resource pdata.Resource, metric pdata.Metric) bool {
-	metricIdentifier := dataPointIdentifier(resource, metric, dp.LabelsMap())
-
-	start := nsp.history[metricIdentifier]
-	// If this is the first time we've observed this unique metric,
-	// record it as the start point and do not report this data point
-	if start == nil {
-		dps := metric.DoubleSum().DataPoints()
-		newDP := pdata.NewDoubleDataPoint()
-		dps.At(0).CopyTo(newDP)
-
-		newStart := startPoint{
-			dataType:        pdata.MetricDataTypeDoubleSum,
-			doubleDataPoint: &newDP,
-			lastDoubleValue: newDP.Value(),
-		}
-		nsp.history[metricIdentifier] = &newStart
-
-		return false
+	if out.Len() > 0 {
+		out.CopyTo(dps)
+		return true, nil
 	}
-
-	// If this data is older than the start point, we can't meaningfully report this point
-	// TODO - consider resetting on two subsequent data points older than current start timestamp.
-	// This could signify a permanent clock change.
-	if dp.Timestamp() <= start.doubleDataPoint.Timestamp() {
-		nsp.logger.Info(
-			"data point being processed older than last recorded reset, will not be emitted",
-			zap.String("lastRecordedReset", start.doubleDataPoint.Timestamp().String()),
-			zap.String("dataPoint", dp.Timestamp().String()),
-		)
-		return false
-	}
-
-	// If data has rolled over or the counter has been restarted for
-	// any other reason, grab a new start point and do not report this data
-	if dp.Value() < start.lastDoubleValue {
-		dp.CopyTo(*start.doubleDataPoint)
-		start.lastDoubleValue = dp.Value()
-
-		return false
-	}
-
-	start.lastDoubleValue = dp.Value()
-	dp.SetValue(dp.Value() - start.doubleDataPoint.Value())
-	dp.SetStartTimestamp(start.doubleDataPoint.Timestamp())
-
-	return true
+	return false, nil
```

Comment:
Yeah, I think you're right that it could be. Because of the refactoring I removed the redundant check of the metric's type that was here, so there's no longer any error conditions. I didn't change the function signature to keep the diff smaller, but by removing this I was able to clean up a whole tree of error handling, so done now.